main.go

```diff
@@ -4,57 +4,119 @@ package main
+)
 
+// NVMe spec says there are 0 to 8 temperature sensors
 type nvmeCollector struct {
-	nvmeCriticalWarning *prometheus.Desc
-	nvmeTemperature *prometheus.Desc
-	nvmeAvailSpare *prometheus.Desc
-	nvmeSpareThresh *prometheus.Desc
-	nvmePercentUsed *prometheus.Desc
+	nvmeCriticalWarning                    *prometheus.Desc
+	nvmeAvailableSpare                     *prometheus.Desc
+	nvmeTempThreshold                      *prometheus.Desc
+	nvmeReliabilityDegraded                *prometheus.Desc
+	nvmeRO                                 *prometheus.Desc
+	nvmeVMBUFailed                         *prometheus.Desc
+	nvmePMRRO                              *prometheus.Desc
+	nvmeTemperature                        *prometheus.Desc
+	nvmeAvailSpare                         *prometheus.Desc
+	nvmeSpareThresh                        *prometheus.Desc
+	nvmePercentUsed                        *prometheus.Desc
 	nvmeEnduranceGrpCriticalWarningSummary *prometheus.Desc
-	nvmeDataUnitsRead *prometheus.Desc
-	nvmeDataUnitsWritten *prometheus.Desc
-	nvmeHostReadCommands *prometheus.Desc
-	nvmeHostWriteCommands *prometheus.Desc
-	nvmeControllerBusyTime *prometheus.Desc
-	nvmePowerCycles *prometheus.Desc
-	nvmePowerOnHours *prometheus.Desc
-	nvmeUnsafeShutdowns *prometheus.Desc
-	nvmeMediaErrors *prometheus.Desc
-	nvmeNumErrLogEntries *prometheus.Desc
-	nvmeWarningTempTime *prometheus.Desc
-	nvmeCriticalCompTime *prometheus.Desc
-	nvmeThmTemp1TransCount *prometheus.Desc
-	nvmeThmTemp2TransCount *prometheus.Desc
-	nvmeThmTemp1TotalTime *prometheus.Desc
-	nvmeThmTemp2TotalTime *prometheus.Desc
+	nvmeDataUnitsRead                      *prometheus.Desc
+	nvmeDataUnitsWritten                   *prometheus.Desc
+	nvmeHostReadCommands                   *prometheus.Desc
+	nvmeHostWriteCommands                  *prometheus.Desc
+	nvmeControllerBusyTime                 *prometheus.Desc
+	nvmePowerCycles                        *prometheus.Desc
+	nvmePowerOnHours                       *prometheus.Desc
+	nvmeUnsafeShutdowns                    *prometheus.Desc
+	nvmeMediaErrors                        *prometheus.Desc
+	nvmeNumErrLogEntries                   *prometheus.Desc
+	nvmeWarningTempTime                    *prometheus.Desc
+	nvmeCriticalCompTime                   *prometheus.Desc
+	nvmeTemperatureSensors                 []*prometheus.Desc
+	nvmeThmTemp1TransCount                 *prometheus.Desc
+	nvmeThmTemp2TransCount                 *prometheus.Desc
+	nvmeThmTemp1TotalTime                  *prometheus.Desc
+	nvmeThmTemp2TotalTime                  *prometheus.Desc
+	temperatureScale                       *string
 }
 
 // nvme smart-log field descriptions can be found on page 180 of:
 // https://nvmexpress.org/wp-content/uploads/NVM-Express-Base-Specification-2_0-2021.06.02-Ratified-5.pdf
 
-func newNvmeCollector() prometheus.Collector {
+func newNvmeCollector(temperatureScale *string) prometheus.Collector {
+	var sensorDescriptions []*prometheus.Desc
```

Comment:
Since you have maxTempSensors, could you make() this slice with a defined capacity? It would be much more performant than a dynamically resized slice.

Reply:
2 nanoseconds ?